pkg/network/server.go

```diff
@@ -826,11 +830,66 @@ func (s *Server) RelayTxn(t *transaction.Transaction) RelayReason {
 
 // broadcastTX broadcasts an inventory message about new transaction.
 func (s *Server) broadcastTX(t *transaction.Transaction) {
-	msg := s.MkMsg(CMDInv, payload.NewInventory(payload.TXType, []util.Uint256{t.Hash()}))
+	select {
+	case s.transactions <- t:
+	case <-s.quit:
+	}
+}
+
+func (s *Server) broadcastTxHashes(hs []util.Uint256) {
+	msg := s.MkMsg(CMDInv, payload.NewInventory(payload.TXType, hs))
 
 	// We need to filter out non-relaying nodes, so plain broadcast
 	// functions don't fit here.
 	s.iteratePeersWithSendMsg(msg, Peer.EnqueuePacket, func(p Peer) bool {
 		return p.Handshaked() && p.Version().Relay
 	})
 }
+
+// broadcastTxLoop is a loop for batching and sending
+// transactions hashes in and INV payload.
+func (s *Server) broadcastTxLoop() {
+	const (
+		batchTime = time.Millisecond * 50
+		batchSize = 32
+	)
+
+	txs := make([]util.Uint256, 0, batchSize)
+
+	timer := time.NewTimer(batchTime)
+	defer timer.Stop()
+
+	broadcast := func() {
+		go s.broadcastTxHashes(txs)
+		txs = make([]util.Uint256, 0, batchSize)
+
+		timer.Stop()
+		timer = time.NewTimer(batchTime)
+	}
+
+	for {
+		select {
+		case <-s.quit:
+			return
+		case <-timer.C:
+			if len(txs) > 0 {
+				broadcast()
+			}
+		default:
```

Comment:
I don't think we need this select, it's not a problem to miss a timer if there are incoming transactions (we'll send a batch if it's full anyway).